Add support for child base route pattern and name in admin class

ksom opened the pull request:
Hello,

This is a rebase of PR #3095 from 2.3 to master.
https://github.com/sonata-project/SonataAdminBundle/pull/3095

Copy of previous PR documentation :
------------------------------------------

I have tried to set the `$baseRoutePattern` in a child admin class and the behavior is that the parent base route pattern prefix is overridden by the child base route pattern.

Example:
```php
// ParentAdmin.php
protected $baseRoutePattern = 'my-parent';

// ChildAdmin.php
protected $baseRoutePattern = 'my-child';
```

So currently the generated route is : `/admin/my-child/list` (which is generated twice: once for the generic admin and once for the child (parent specific) admin)

This prevents filtering the children of my `Parent` instance.

With this pull request the generated route will be : `/admin/my-parent/{id}/my-child/list`

I did the same behavior with base route name.

I didn't found any tests that check the behavior when a custom base route name or a custom baseRoutePattern.

There is no setter to set the baseRoutePattern or the baseRouteName so it's hard to test, and, I think, out of the scope of this PR.

Do you accept this merge request as-is and after I do a new merge request for tests ?
Would you consider this PR acceptable as-is and I could, write comprehensive tests for explicitly specifying them, later on.

Thanks with advance.

Virgile

ksom commented:
@Soullivaneuh 
I rebased my branch on master, as you asked me.

OskarStark commented:
i would recommend to add ```setCachedBaseRoute....```. and ```getCached...```

ksom commented:
@OskarStark 
They're internal cached properties, I don't understand why you recommend to add these.

OskarStark commented:
i would say it's something like best practice, isn't it?

i like these as it is more easy to stay BC in future....
just saying

ksom commented:
Sorry, but I think in this case it's not relevant.

2015-09-16 10:35 GMT-04:00 Oskar Stark <notifications@github.com>:

> i would say it's something like best practice, isn't it?
>
> i like these as it is more easy to stay BC in future....
> just saying
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/sonata-project/SonataAdminBundle/pull/3250#issuecomment-140760543>
> .
>

jvasseur commented:
cachedBaseRouteName and cachedBaseRoutePattern should probably be private instead of protected since they should only be used as cache for the base admin class.

It would help keeping BC in the future.

OskarStark commented:
:+1:

Am Mittwoch, 16. September 2015 schrieb Jérôme :

> cachedBaseRouteName and cachedBaseRoutePattern should probably be private
> instead of protected since they should only be used as cache for the base
> admin class.
>
> It would help keeping BC in the future.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/sonata-project/SonataAdminBundle/pull/3250#issuecomment-140766041>
> .
>

stevedavid commented:
:+1:

Soullivaneuh commented:
Could you please update the documentation?

ksom commented:
Sorry, I have no time at the moment. I will do theses changes as soon as possible.

chartinger commented:
any updates on this?

ksom commented:
@chartinger Will be updated this week end.

OskarStark commented:
Please don't foget the docs :+1:

ksom commented:
@Soullivaneuh @OskarStark @chartinger @jvasseur 

Hi,

This PR is updated, I switched cached properties from protected to private and updated the documentation for routing.

OskarStark commented:
Thank you @ksom 

Can you please add some tests?

ksom commented:
Thank you for your response.
There is no setter to set the baseRoutePattern or the baseRouteName so it's hard to test, and, I think, out of the scope of this PR.

OskarStark commented:
but if you would create 2 classes and set the member variables, the route must be generated correct, IMO. This should be easy to test, shouldn't it?

ksom commented:
@OskarStark This will be done this week end.

OskarStark commented:
great to hear @ksom 

please rebase your PR to fix travis, thank you!

ksom commented:
@OskarStark @Soullivaneuh 
:+1: 
Updated this PR with tests.

OskarStark commented:
great work :+1: 

can you please squash your commits @ksom ?

@rande can we merge this PR?

ksom commented:
Yes I can, do you want only one commit ? @OskarStark

OskarStark commented:
Yes please one commit

Am Samstag, 12. Dezember 2015 schrieb ksom :

> Yes I can, do you want only one commit ? @OskarStark
> <https://github.com/OskarStark>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/sonata-project/SonataAdminBundle/pull/3250#issuecomment-164171746>
> .
>

ksom commented:
@OskarStark Done, thanks.

OskarStark commented:
Thank you @ksom, great work :+1:

OskarStark commented:
can you please check your application with the current master and check if everything is working?

thank you
